### ISSUE TITLE
How are you auto-casting with just the = sign

### ISSUE THREAD

**Issue by eonist** [seivan/ScalarArithmetic#17]
IN your sample code you have: 

```swift
cgFloatValue = doubleValue
doubleValue = intValue
doubleValue = cgFloatValue
```

But I can't find the auto infer code in your src. Some sort of extension or?

**Comment by seivan** [comment]
@eonist Sorry, it looks like the `Readme` is a bit dated. I was relying on the `@conversion` keyword pre Swift 2 and it seems it's gone. 
It allowed for implicit conversion on assignment. 

The only thing possible today, is what's defined in the test-suite(s).
I'll fix the Readme once I migrate over to 3.0. 

Sorry for the inconvenience.

**Comment by eonist** [comment]
So something like this:

``` swift
let someDouble:Double = 4
let someFloat:CGFloat = someDouble
```

Won't work in swift 3.0?

**Comment by seivan** [comment]
Not without some Arithmetic operation, e.g addition. 
let someFloat:CGFloat = 0 + someInt
or 
let someFloat:CGFloat = otherFloat + someInt
or
someFloat += someInt

Since we don’t got implicit conversion on assignment, the next best thing is implicit conversion through operators.

**Comment by eonist** [comment]
Right. Would be so awesome to not worry about casting Numerical types. I guess one could make an universal cast method by using inference and generics. like:

``` swift
let someFloat:Float = 4
let someInt:Int = someFloat.cast()
let someDouble:Double = someFloat.cast()
let someCGFloat:CGFloat = someFloat.cast()

extension Float{
    func <T>cast()->T{
        if(T is Int){return Int(self)}
        else if(T is Double){return Double(self)}
        else if(T is CGFloat){return CGFloat(self)}
        else {fatalError("numerical type not supported")}
    }
}
```

Better yet. Make a protocol with an extension that every numerical type could extend. So that you only would have to write the extension once. extension Float:NumericalConversional... and extension CGFloat:NumericalConversional...etc

**Comment by seivan** [comment]
That’s actually been on my mind. But not sure about the implications.
So far I’ve resorted to + 0 just to be explicit about it. But maybe cast() is more explicit.

**Comment by eonist** [comment]
What i personally do Is:

``` swift
let someCGFloat:CGFloat = 4
let someInt:Int = someCGFloat.int
let someDouble:Int = someCGFloat.double
let someString:String = someCGFloat.string
```

But this solution has code all over the place. But the suggested cast solution would be more centralised and easier to reason with. Thats why I stared your project , I tought you had found a way to infer types with the "=" sign. :)

**Comment by seivan** [comment]
I am sorry, I did in Swift 1.0. But they sorta removed it, and I understand since implicit casts could be dangerous.
I actually like your way better, but a `casted` property could work without the huge switch case you had in mind if you rely on types. I could spike up a quick idea if you’d want that.

**Comment by eonist** [comment]
Yes, definitely avoid the huge if else clause with polymorphism. And just extend every Numerical type instead. If its all in one .swift class its easy enough to reason with.

**Comment by eonist** [comment]
If you have arrays with numbers you can also do something like this:

``` swift
let someFloats:[Float] = [1,2,3,4]
let someCGFloats:[CGFloat] = someFloats.cast()

extension Array{
   func cast<T:NumericConversional>() -> [T]{
      return self.map { $0.cast() }
   }
}

```

NOTE: NumericConversional would need to be a protocol that you add when you make the extension for each Number type. etc etc

**Comment by seivan** [comment]
For Sequence extensions  - You'd probably want to use  _either_ a flatMap or rethrow errors.

**Comment by eonist** [comment]
Isn't flatmap used when a value can be nil? Are you thinking about numerical values that can be NaN? I can't remember but I think Int or UInt can't be NaN

**Comment by seivan** [comment]
Yep, not to mention between unsigned and signed and such. I recall seeing optional inits. But I might be mistaken.

**Comment by eonist** [comment]
unsigned and signed == special cases no? 

<img width="320" alt="img" src="https://dl.dropboxusercontent.com/u/2559476/Screen Shot 2016-10-11 at 15.21.52.png">

Anyways I usually just write for the simplest case first, if something needs more features then one could method overload... like for cases where the array items can be nil:

``` swift
 func cast<T?:NumericConversional>() -> [T?]{
   return self.flatMap { $0.cast() }
 }
```

Also i'm not sure if flatMap handles NaN values. asserting for isNaN could be a substitute.

**Comment by seivan** [comment]
flatMap doesnt out of the box, but you'd have assertions in there on whatever protocol you have.
